src/types/tuple.rs

```diff
@@ -535,6 +537,109 @@ fn type_output() -> TypeInfo {
             TypeInfo::Tuple(Some(vec![$( $T::type_output() ),+]))
         }
+
+        #[inline]
+        fn __py_call_vectorcall1<'py>(
+            self,
+            py: Python<'py>,
+            function: Borrowed<'_, 'py, PyAny>,
+        ) -> PyResult<Bound<'py, PyAny>> {
+            cfg_if::cfg_if! {
+                if #[cfg(all(Py_3_9, not(any(PyPy, GraalPy, Py_LIMITED_API))))] {
+                    // We need this to drop the arguments correctly.
+                    let args_bound = [$(self.$n.into_py(py).into_bound(py),)*];
+                    if $length == 1 {
+                        unsafe {
+                            ffi::PyObject_CallOneArg(function.as_ptr(), args_bound[0].as_ptr()).assume_owned_or_err(py)
+                        }
+                    } else {
+                        // Prepend one null argument for `PY_VECTORCALL_ARGUMENTS_OFFSET`.
+                        let mut args = [std::ptr::null_mut(), $(args_bound[$n].as_ptr()),*];
+                        unsafe {
+                            ffi::PyObject_Vectorcall(
+                                function.as_ptr(),
+                                args.as_mut_ptr().add(1),
+                                $length + ffi::PY_VECTORCALL_ARGUMENTS_OFFSET,
+                                std::ptr::null_mut(),
+                            )
+                            .assume_owned_or_err(py)
+                        }
+                    }
+                } else {
+                    function.call1(<Self as IntoPy<Py<PyTuple>>>::into_py(self, py).into_bound(py))
+                }
+            }
+        }
+
+        #[inline]
+        fn __py_call_vectorcall<'py>(
+            self,
+            py: Python<'py>,
+            function: Borrowed<'_, 'py, PyAny>,
+            kwargs: Option<Borrowed<'_, '_, PyDict>>,
+        ) -> PyResult<Bound<'py, PyAny>> {
+            cfg_if::cfg_if! {
+                if #[cfg(all(Py_3_9, not(any(PyPy, GraalPy, Py_LIMITED_API))))] {
+                    match kwargs {
+                        Some(kwargs) => unsafe {
+                            // We need this to drop the arguments correctly.
+                            let args_bound = [$(self.$n.into_py(py).into_bound(py),)*];
+                            // Prepend one null argument for `PY_VECTORCALL_ARGUMENTS_OFFSET`.
+                            let mut args = [std::ptr::null_mut(), $(args_bound[$n].as_ptr()),*];
+                            ffi::PyObject_VectorcallDict(
+                                function.as_ptr(),
+                                args.as_mut_ptr().add(1),
+                                $length + ffi::PY_VECTORCALL_ARGUMENTS_OFFSET,
+                                kwargs.as_ptr(),
+                            )
+                            .assume_owned_or_err(py)
+                        },
+                        None => <Self as IntoPy<Py<PyTuple>>>::__py_call_vectorcall1(self, py, function),
```

Comment:
I think it's legal to call `PyObject_VectorcallDict` with a null ptr and let the interpreter decide how to optimise; I think that'd keep the code simpler imo.

Reply:
I think it make sense to define a `Option<Borrowed/&Bound>::as_ptr_or_null()`. It is commonly used. But this can be done later.